hazelcast/src/main/java/com/hazelcast/query/impl/GlobalIndexPartitionTracker.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.query.impl;
+
+import com.hazelcast.internal.util.collection.PartitionIdSet;
+
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Tracker of indexed partition for global indexes.
+ * <p>
+ * Used by {@link InternalIndex} to keep track of indexed partitions and decide whether the query
+ * using this index could be executed or not.
+ */
+public class GlobalIndexPartitionTracker {
+
+    private static final long VERSION_INITIAL = 0;
+
+    /** Lock to serialize updates to the state. */
+    private final ReentrantLock lock = new ReentrantLock();
+
+    /** Number of partitions in the cluster. */
+    private final int partitionCount;
+
+    /** Current state. */
+    private final AtomicReference<State> state;
```

Comment:
The problem with `volatile State` is that it cannot be initialized in the constructor safely, because formally the assignment of the constructed tracker could be moved before the assignment of the `state` field, leading to NPE. The solution to this is either some `final` object holder, like in the current PR, or some sort of additional null-checks in several places. I tried the latter, and it appeared more complex than just `final` holder.